File: Assets/Scripts/Services/GitHubService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Newtonsoft.Json;
using UnityEngine;
using UnityEngine.Networking;
using GitVisualizer.Models;

namespace GitVisualizer.Services
{
    public class GitHubService
    {
        private const string ApiBaseUrl = "https://api.github.com";
        private const string AcceptHeader = "application/vnd.github+json";
        private const string ApiVersionHeader = "2022-11-28";
        private const int CommitsPerBranch = 50;
        private const int MaxBranchesPerPage = 100;

        public class RepoDataResult
        {
            public Repository Repository;
            public Branch[] Branches;
            public Dictionary<string, Commit[]> CommitsByBranch = new Dictionary<string, Commit[]>();
        }

        public async Task<RepoDataResult> FetchRepoData(string owner, string repoName, string personalAccessToken)
        {
            if (string.IsNullOrWhiteSpace(owner) || string.IsNullOrWhiteSpace(repoName))
            {
                Debug.LogError("[GitHubService] Owner and repo name are required.");
                return null;
            }

            var result = new RepoDataResult();

            try
            {
                result.Repository = await FetchRepositoryAsync(owner, repoName, personalAccessToken);
                if (result.Repository == null) return null;

                result.Branches = await FetchBranchesAsync(owner, repoName, personalAccessToken);
                if (result.Branches == null) return null;

                foreach (var branch in result.Branches)
                {
                    var commits = await FetchCommitsAsync(owner, repoName, branch.name, personalAccessToken);
                    if (commits != null)
                        result.CommitsByBranch[branch.name] = commits;
                }

                return result;
            }
            catch (Exception ex)
            {
                Debug.LogError($"[GitHubService] {ex.Message}\n{ex.StackTrace}");
                return null;
            }
        }

        private async Task<Repository> FetchRepositoryAsync(string owner, string repoName, string token)
        {
            var json = await SendGetRequestAsync($"{ApiBaseUrl}/repos/{owner}/{repoName}", token);
            if (json == null) return null;
            try { return JsonConvert.DeserializeObject<Repository>(json); }
            catch (JsonException ex) { Debug.LogError($"[GitHubService] Parse repo: {ex.Message}"); return null; }
        }

        private async Task<Branch[]> FetchBranchesAsync(string owner, string repoName, string token)
        {
            var url = $"{ApiBaseUrl}/repos/{owner}/{repoName}/branches?per_page={MaxBranchesPerPage}";
            var json = await SendGetRequestAsync(url, token);
            if (json == null) return null;
            try { return JsonConvert.DeserializeObject<Branch[]>(json) ?? Array.Empty<Branch>(); }
            catch (JsonException ex) { Debug.LogError($"[GitHubService] Parse branches: {ex.Message}"); return null; }
        }

        private async Task<Commit[]> FetchCommitsAsync(string owner, string repoName, string branchShaOrName, string token)
        {
            var encoded = Uri.EscapeDataString(branchShaOrName);
            var url = $"{ApiBaseUrl}/repos/{owner}/{repoName}/commits?sha={encoded}&per_page={CommitsPerBranch}";
            var json = await SendGetRequestAsync(url, token);
            if (json == null) return null;
            try { return JsonConvert.DeserializeObject<Commit[]>(json) ?? Array.Empty<Commit>(); }
            catch (JsonException ex) { Debug.LogError($"[GitHubService] Parse commits: {ex.Message}"); return null; }
        }

        private async Task<string> SendGetRequestAsync(string url, string token)
        {
            using (var request = UnityWebRequest.Get(url))
            {
                request.SetRequestHeader("Accept", AcceptHeader);
                request.SetRequestHeader("X-GitHub-Api-Version", ApiVersionHeader);
                if (!string.IsNullOrEmpty(token))
                    request.SetRequestHeader("Authorization", $"Bearer {token}");

                var operation = request.SendWebRequest();
                while (!operation.isDone)
                    await Task.Yield();

                if (request.result == UnityWebRequest.Result.Success)
                    return request.downloadHandler.text;

                HandleRequestError(request, url);
                return null;
            }
        }

        private void HandleRequestError(UnityWebRequest request, string url)
        {
            var code = request.responseCode;
            var msg = request.error ?? request.downloadHandler?.text ?? "Unknown error";

            if (code == 403)
                Debug.LogError("[GitHubService] Rate limit (403). Use a PAT for higher limits.");
            else if (code == 404)
                Debug.LogError($"[GitHubService] Not found (404): {url}");
            else
                Debug.LogError($"[GitHubService] {code}: {url} - {msg}");
        }
    }
}

File: src/main/java/com/gitvisualizer/service/GitHubService.java
package com.gitvisualizer.service;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.gitvisualizer.graph.GraphVisualizationBuilder;
import com.gitvisualizer.model.Branch;
import com.gitvisualizer.model.Commit;
import com.gitvisualizer.model.Repository;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

/** Fetch repository metadata, branches, and recent commits via the GitHub REST API. */
public class GitHubService {

    private static final String DEFAULT_API = "https://api.github.com";
    private static final String ACCEPT = "application/vnd.github+json";
    private static final String API_VERSION = "2022-11-28";

    private final HttpClient httpClient;
    private final ObjectMapper mapper;
    private final String apiBaseUrl;

    public GitHubService() {
        this(DEFAULT_API, HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build());
    }

    public GitHubService(String apiBaseUrl, HttpClient httpClient) {
        this.apiBaseUrl = Objects.requireNonNullElse(apiBaseUrl, DEFAULT_API).replaceAll("/$", "");
        this.httpClient = httpClient != null ? httpClient : HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build();
        this.mapper = new ObjectMapper()
                .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    public record RepoDataResult(
            Repository repository,
            Branch[] branches,
            Map<String, Commit[]> commitsByBranch
    ) {
    }

    public RepoDataResult fetchRepoData(String owner, String repoName, String personalAccessToken)
            throws IOException, InterruptedException {
        if (owner == null || owner.isBlank() || repoName == null || repoName.isBlank()) {
            return null;
        }
        Repository repo = fetchRepository(owner, repoName, personalAccessToken);
        if (repo == null) {
            return null;
        }
        Branch[] branches = fetchBranches(owner, repoName, personalAccessToken);
        if (branches == null) {
            return null;
        }
        Map<String, Commit[]> commitsByBranch = new LinkedHashMap<>();
        for (Branch branch : branches) {
            if (branch == null || branch.name() == null) {
                continue;
            }
            Commit[] commits = fetchCommits(owner, repoName, branch.name(), personalAccessToken);
            if (commits != null) {
                commitsByBranch.put(branch.name(), commits);
            }
        }
        return new RepoDataResult(repo, branches, commitsByBranch);
    }

    public Repository fetchRepository(String owner, String repoName, String token)
            throws IOException, InterruptedException {
        String json = get("/repos/" + owner + "/" + repoName, token);
        if (json == null) {
            return null;
        }
        return mapper.readValue(json, Repository.class);
    }

    public Branch[] fetchBranches(String owner, String repoName, String token)
            throws IOException, InterruptedException {
        String url = "/repos/" + owner + "/" + repoName + "/branches?per_page=" + GraphVisualizationBuilder.MAX_BRANCHES_PER_PAGE;
        String json = get(url, token);
        if (json == null) {
            return null;
        }
        Branch[] arr = mapper.readValue(json, Branch[].class);
        return arr != null ? arr : new Branch[0];
    }

    public Commit[] fetchCommits(String owner, String repoName, String branchShaOrName, String token)
            throws IOException, InterruptedException {
        String encoded = URLEncoder.encode(branchShaOrName, StandardCharsets.UTF_8);
        String url = "/repos/" + owner + "/" + repoName + "/commits?sha=" + encoded
                + "&per_page=" + GraphVisualizationBuilder.COMMITS_PER_BRANCH;
        String json = get(url, token);
        if (json == null) {
            return null;
        }
        Commit[] arr = mapper.readValue(json, Commit[].class);
        return arr != null ? arr : new Commit[0];
    }

    private String get(String path, String token) throws IOException, InterruptedException {
        URI uri = URI.create(apiBaseUrl + path);
        HttpRequest.Builder b = HttpRequest.newBuilder(uri)
                .timeout(Duration.ofSeconds(60))
                .header("Accept", ACCEPT)
                .header("X-GitHub-Api-Version", API_VERSION)
                .GET();
        if (token != null && !token.isEmpty()) {
            b.header("Authorization", "Bearer " + token);
        }
        HttpResponse<String> response = httpClient.send(b.build(), HttpResponse.BodyHandlers.ofString());
        int code = response.statusCode();
        if (code >= 200 && code < 300) {
            return response.body();
        }
        return null;
    }

    public ObjectMapper objectMapper() {
        return mapper.copy();
    }
}
